dex/wait/queue_test.go

```diff
@@ -110,17 +121,108 @@ func TestTaperingQueue(t *testing.T) {
+				waiterNumber, len(wantTriesTimes), len(gotTriesTimed))
+		}
+		var (
+			prevWantTryTime time.Time
+			prevGotTryTimed time.Time
+		)
+		for i, wantTryTime := range wantTriesTimes {
+			gotTryTimed := gotTriesTimed[i]
+			if i == 0 {
+				// Can't compare try difference for first try since there is nothing
+				// to compare against.
+				prevWantTryTime = wantTryTime
+				prevGotTryTimed = gotTryTimed
+				continue
+			}
+			// Check that waiter tapering works, in other words each waiter try attempt
+			// doesn't execute sooner than we expect.
+			// We compare the actual observed time difference between two adjacent waiter
+			// try-attempts with synthetically calculated one (in wantTryDiff var), the
+			// actual observed should be higher-or-equal because there is additional
+			// code executing (scheduling/executing retry-attempts and such).
+			wantTryDiff := wantTryTime.Sub(prevWantTryTime)
+			gotTryDiff := gotTryTimed.Sub(prevGotTryTimed)
+			if gotTryDiff < wantTryDiff {
+				t.Fatalf("expected waiter %d to have time difference between tries %d-%d be > %v, "+
+					"got time difference: %v", waiterNumber, i, i-1, wantTryDiff, gotTryDiff)
+			}
+			prevWantTryTime = wantTryTime
+			prevGotTryTimed = gotTryTimed
+		}
+	}
+}
+
+func Test_nextTick(t *testing.T) {
+	const fastestInterval, slowestInterval = 100 * time.Millisecond, 500 * time.Millisecond
+	var gotTicks []time.Time
+	now := time.Now()
+	expiration := now.Add(time.Hour)
+
+	// First tick happens right away.
+	gotTicks = append(gotTicks, now)
+	for tick := 1; tick <= 19; tick++ {
+		gotTicks = append(gotTicks, nextTick(tick, slowestInterval, fastestInterval,
+			gotTicks[tick-1], expiration))
+	}
+	wantTicks := expTickSchedule(now, fastestInterval, slowestInterval)
 
-	for i := range resultOrder {
-		if resultOrder[i] != expOrder[i] {
-			t.Fatalf("wrong result order expected %+x, got %+x", expOrder, resultOrder)
+	// To check expiration on the last tick.
+	gotTicks = append(gotTicks, nextTick(20, slowestInterval, fastestInterval,
+		expiration, expiration))
+	wantTicks[len(wantTicks)-1] = expiration
+
+	for i, want := range wantTicks {
+		got := gotTicks[i]
+		if want != got {
+			t.Fatalf("expected tick %d to be: %v, got: %v", i, want, got)
 		}
 	}
 }
+
+// expTickSchedule returns expected tick schedule with a certain startTime.
+func expTickSchedule(startTime time.Time, fastestInterval, slowestInterval time.Duration) []time.Time {
+	expectedTicks := [21]time.Time{} // 21 element should be enough for all our needs in these tests.
+	expectedTicks[0] = startTime
+	expectedTicks[1] = expectedTicks[0].Add(fastestInterval)
+	expectedTicks[2] = expectedTicks[1].Add(fastestInterval)
+	expectedTicks[3] = expectedTicks[2].Add(fastestInterval + time.Duration(math.Round((float64(1)/12)*float64(slowestInterval-fastestInterval))))
+	expectedTicks[4] = expectedTicks[3].Add(fastestInterval + time.Duration(math.Round((float64(2)/12)*float64(slowestInterval-fastestInterval))))
```

Comment:
So these 12s are fullyTapered (15) - fullSpeedTicks (3)?

Reply:
Yes, updated to show this explicitly.